### ISSUE TITLE
How to set timeout in restart role Master?

### ISSUE THREAD

**Issue by ARogovsky** [dalibo/PAF#14]
Hi!
I have a very large DB with diffirent countrys. Due by bandwith limitations every operation have timeout 2mins.
I have issue with very fast Master switching. I need have option to set timeout for switch Master.
Standart ocf was supported this option:
`op monitor interval=30s on-fail=restart role=Master timeout=160s`
PAF does not support it. Can you suggest me something?

**Comment by mrimbault** [comment]
Hi,

I'm not sure to understand what the issue is.
The "timeout" option you're considering is not part of the agent code, but part of the resource declaration.
The agent itself does not deal with timeouts, this is part of good coding practice when developing a Resource Agent (see http://www.linux-ha.org/doc/dev-guides/_timeouts.html).

In consequence, the "timeout" option used when creating or updating a PAF resource should work as intended.
For instance, with a configured 160s timeout, the following monitor action (I hardcoded a `sleep 150` in the agent just for testing) does not fail (exit code "8" means "running master"):

    Apr 12 14:32:53 [3053] hanode1       lrmd:    debug: log_finished:      finished - rsc:pgsqld action:monitor call_id:101 pid:22624 exit-code:8 exec-time:150371ms queue-time:0ms

But with a hardcoded `sleep 300`, we got the expected timeout:

    Apr 12 14:37:16 [3053] hanode1       lrmd:  warning: child_timeout_callback:    pgsqld_monitor_9000 process (PID 24351) timed out
    Apr 12 14:37:16 [3053] hanode1       lrmd:  warning: operation_finished:        pgsqld_monitor_9000:24351 - timed out after 160000ms

We can also check the timeout is effectively written in the resource configuration this way:

    # crm_resource -r pgsqld -q | grep monitor
        <op id="pgsqld-monitor-interval-9s" interval="9s" name="monitor" on-fail="restart" role="Master" timeout="160s"/>
        <op id="pgsqld-monitor-interval-10s" interval="10s" name="monitor" on-fail="restart" role="Slave" timeout="160s"/>

If the timeout parameter is still not working for you, could you provide us with any error message you get when trying to set it ?

What also puzzles me in this issue is that you're considering this parameter in relation with bandwith limitations.
The resource timeout configuration is mostly useful for local actions, not really for network bandwidth and latency between nodes considerations.
The most easy way to take in account a slow responding or otherwise unreliable network is to adapt the "migration-threshold" and "failure-timeout' cluster parameters.
You could also check the corosync's token timeout configuration, but I don't recommend this as this may results in Pacemaker misbehaving.

Finally, you have to keep in mind that, in any high availability configuration, reliability is the most important thing.
So if the network is really not reliable, it will seriously impair your high availability cluster, and there's not many things you can do with Pacemaker alone.
You may want to check the section of the documentation dealing with multi-site clusters: http://clusterlabs.org/doc/en-US/Pacemaker/1.1/html/Pacemaker_Explained/ch15.html

**Comment by ARogovsky** [comment]
Well, my task is very simple.
If Master server is lost, RAF should waiting 2mins until switch new Master.
Now RAF switch Master immediatley.

I was try apply:
`op monitor interval=30s on-fail=restart role=Master timeout=160s`
And get:
`WARNING: pgsqld: action monitor_Master not advertised in meta-data, it may not be supported by the RA`

**Comment by mrimbault** [comment]
Thanks for the details.
About your warning message, it seems to me that it may be related to the resource configuration using `crmsh`.
I usually use `pcs` so I'm not completely sure, I will try to reproduce it.
If that's possible, could you provide the complete resource configuration command ?

Could you also clarify what kind of situation do you want to deal with when you say you want the switch to wait when the Master server is "lost".
I can see several very different cases:
  * the PostgreSQL primary instance is down: the monitor action will usually immediately fail (regardless of the configured timeout), and depending on the returned error code ("soft" or "hard"), a recover of the resource ("stop" / "start") will be attempted, or a failover will be immediately be triggered ;
    * if this is a "soft" error, the monitor's configured interval and the value of the "migration-threshold" parameter will determine how much time before a failover is triggered
    * for exemple, with a monitor interval of "30s" and a "migration-threshold" of "5", a failover will be triggered at the fifth "soft" error, so about two minutes after the first failure
  * the PostgreSQL primary instance is not responsive to the local "monitor" action: the resource configured timeout will determine how much time before a "soft" error is returned to the cluster ;
    * this will only occur if the "monitor" action is very slow to proceed, which is usually not the case (all stuff done by the PAF RA for a "monitor" action is local to the node, so it should not be affected by the network bandwidth or latency)
  * the node hosting the PostgreSQL primary instance is completely down or not responsive (this can be due to a slow responding network): corosync will detect that based on its token timeout, and Pacemaker will then immediately trigger a fencing+Master role failover.

In short, there is difference between:
  * a resource that is really down ;
  * a resource that is not responsive to the "monitor" action;
  * a node that is not responsive.

The appropriate configuration depends on the situation you're in.

**Comment by ARogovsky** [comment]
This is my config:
```
crm(live)configure# show
node 1084754433: a.server \
	attributes pgsql-data-status="STREAMING|ASYNC" maintenance=off
node 1084754434: b.server \
	attributes pgsql-data-status=LATEST maintenance=off standby=off
node 1084754435: c.server \
	attributes pgsql-data-status="STREAMING|ASYNC" maintenance=off
primitive pgsql-master-ip IPaddr2 \
	params ip=192.168.10.200 nic=peervpn0 \
	op start interval=0s on-fail=restart timeout=60s \
	op monitor interval=10s on-fail=restart timeout=60s \
	op stop interval=0s on-fail=block timeout=60s \
	meta target-role=Started
primitive pgsqld pgsqlms \
	params bindir="/usr/lib/postgresql/9.4/bin" pgdata="/var/lib/postgresql/9.4/main" pghost="/var/run/postgresql" primary_node=a.server recovery_template="/etc/postgresql/9.4/main/recovery.conf.pcmk" start_opts="-c config_file=/etc/postgresql/9.4/main/postgresql.conf" \
	op start interval=0 on-fail=restart timeout=160s \
	op monitor interval=40s on-fail=restart timeout=160s \
	op promote interval=0s on-fail=restart timeout=160s \
	op demote interval=0s on-fail=stop timeout=160s \
	op monitor interval=30s on-fail=restart role=Master timeout=160s \
	op stop interval=0 on-fail=block timeout=160s \
	op notify interval=0s timeout=160s
group master pgsql-master-ip
ms msPostgresql pgsqld \
	meta master-max=1 master-node-max=1 clone-max=3 clone-node-max=1 target-role=Master notify=true is-managed=true
location pgsql-prefers-a msPostgresql role=Master 1000: a.server
colocation set_ip inf: master msPostgresql:Master
order ip_down 0: msPostgresql:demote master:stop symmetrical=false
order ip_up 0: msPostgresql:promote master:start symmetrical=false
property cib-bootstrap-options: \
	dc-version=1.1.12-cdf310a \
	cluster-infrastructure=corosync \
	expected-quorum-votes=3 \
	stonith-enabled=false \
	crmd-transition-delay=0 \
	last-lrm-refresh=1460369932 \
	dc-deadtime=2min \
	default-action-timeout=2min \
	have-watchdog=false
rsc_defaults rsc-options: \
	resource-stickiness=10000 \
	migration-threshold=10

```
I need wait 2mins becouse sometimes network is lag and server stop view for short time. I perfer wait 2 mins before change master.

**Comment by mrimbault** [comment]
I just tested with a Debian 8 cluster configured with `crmsh`, and I had the same Warning message that you when configuring the PostgreSQL resource:

    WARNING: pgsqld: action monitor_Master not advertised in meta-data, it may not be supported by the RA

In spite of this message, the configured timeout for the `monitor` action works as intended:

    Failed Actions:
    * pgsqld_monitor_9000 on hanode2 'unknown error' (1): call=211, status=Timed Out, exitreason='none',
        last-rc-change='Wed Apr 13 14:59:48 2016', queued=0ms, exec=160004ms

I also did not found any reference to a "monitor_Master" or a "monitor_Slave" actions in Clusterlabs and Linux-HA documentations, nor is there any reference to this on existing agents or Pacemaker source code.
It seems to me that the Warning message is a bug in `crmsh`, but as far as I could test, a harmless one.

I don't think there is a problem with PAF RA here, so I will close this issue.
I may have overlooked something, if so please reopen it.

Still, I have several remarks about your configuration, none of them related to PAF.

About your requirement to wait before a failover, you should be OK as you are. With your configuration Pacemaker should wait for at least 4 minutes and 30 seconds between the first "soft" error detected on a monitor action, and a forced failover: your monitor's interval is "30s" and your "migration-threshold" is "10".
You should also add a "failure-timeout" configuration, to clean old failcounts after a certain time.

Obviously if the error encountered by the monitor action is a "hard" one, a failover would be immediate, as it should.

Now, as I previously wrote, the monitor timeout and migration-threshold configuration is all about local failure detection.
If your network is slow or unreliable, it is quite possible that corosync itself will report that the node is not available.
As I mentioned before, you could try to adapt the corosync's token timeout so the node will not be marked "failed" too soon.

Nevertheless, having an unreliable network between your nodes will most certainly cause problems.
Please note that most of the available configuration for Pacemaker is useful for "local" clusters only.
If you're dealing with multiple, geographically dispersed sites, you should probably look at [the related Pacemaker's documentation] (http://clusterlabs.org/doc/en-US/Pacemaker/1.1/html/Pacemaker_Explained/ch15.html).

Note: I see that you explicitly disabled fencing in your configuration ("stonith-enabled=false").
If this configuration is designed for a production system, that's a bad idea, a reliable fencing method is a mandatory element of any high availability configuration.

**Comment by ARogovsky** [comment]
Hi!
Thanks for advices. I will tryi implement your solutions. 
I was used config from http://clusterlabs.org/wiki/PgSQL_Replicated_Cluster, it is have op monitor timeout="60s" interval="3s"  on-fail="restart" role="Master" and this supported by standart pgsql resource agent from pacemaker 1.1.
So, it is depends on resource-agent. This is actions from standart pgsl:
```
<actions>
<action name="start" timeout="120" />
<action name="stop" timeout="120" />
<action name="status" timeout="60" />
<action name="monitor" depth="0" timeout="30" interval="30"/>
<action name="monitor" depth="0" timeout="30" interval="29" role="Master" />
<action name="promote" timeout="120" />
<action name="demote" timeout="120" />
<action name="notify"   timeout="90" />
<action name="meta-data" timeout="5" />
<action name="validate-all" timeout="5" />
<action name="methods" timeout="5" />
</actions>
```

About fencing. I have few servers in diffirent DC. So, I did not have any fenced devices, except suicide and meatware. I perfer just stop services on node which lost quorum and hold services down until operators check status and cleanup counters manual.
May be is this not good idea?

**Comment by mrimbault** [comment]
Hi,

Thanks for the pointer, indeed the Warning message was due to the missing `role="Master"` and `role="Slave"` action lines in the meta-data, I just fixed this in the last commit.
But as already stated, this message was harmless as the configured timeout was taken in account in spite of it.

Thanks for the report!

For your question about fencing and multi-site replication, I think the documentation links I gave you in my previous comment is a good place to start. The "ticket" and "dead man dependency" mechanisms could be what you're looking for.
I don't think this issue is the right place to continue a discussion about this point, but feel free to start a thread about it on the [Pacemaker mailing lists] (http://clusterlabs.org/wiki/Mailing_lists), most probably "users".

**Comment by ARogovsky** [comment]
Thanks for adivces!
I try implement fencing by stonith suicide, but it not worsk - when I disable networking on one node - it is not self rebooted. I don't know why.
May be I will movie fencing on external level - ping hosts via monit, if both host is down - then reboot.
